### ISSUE TITLE
which fold is used to create the model?

### ISSUE THREAD

**Issue by JianWang2016** [pycaret/pycaret#910]
Hi,  Great python package!

I have a couple of questions:

1) The following function fits 5 folds and the accuracy scores (along with other metric scores) are generated for each of the 5 folds.  There are 5 independent models from the cross-validation. I am trying to understand which fold the constructed "myknn" comes from, the one with the best score?
 
myknn = create_model('knn', fold=5)

2) I couldn't find if create_model supports multiple repeats of the cross validation,  as n_repeats in sklearn: cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=100, random_state=1)

3) Besides the summary metrics from the function output, is there a way to get scores from each fold?  It would be useful to know the variance of these scores.

4) Related question in the tuned_dt = tune_model(dt, fold=5, optimize = 'AUC'), for example.  I know the train dataset is used here for the tune_model.  But in the model "tuned_dt", is the input data from one of the 5 folds or as I hope: when the best model is identified, it is trained using the entire X_train and y_train.This is actually related to question 1).

5) Since my dataset is highly unbalanced so I don't use accuracy as scoring metrics.  In the create_model() above, how do I set that up? I use sort='AUC' in compare_models(), optimizer='AUC' in the tune_model() and ensembl_model(), do they mean the same thing, attempting to fit to improve AUC? Why call different names? 
6) in blend_models(method='hard'), all of the output give AUC=0, is this a bug?

Thanks much.

**Comment by Yard1** [comment]
1. Cross-validation (not just in PyCaret, but in general) is only used to determine the mean metrics of all folds in order to reduce the influence of overfitting on the scores. The returned model is fitted on the entire train dataset.

2. You can pass sklearn CV generators in the fold parameter:
```python
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=100, random_state=1)
model = create_model('knn', fold=cv)
```

3. The displayed dataframe shows the scores for each fold. You can get it with `pull()`:
```python
model = create_model('knn', fold=cv)
scores = pull()
```

4. `tune_model` does not use the train dataset at all. Instead, it uses cross-validation and selects the best model by the mean score of all the folds - but the final models are, as in `create_model`, fitted on the entire train set.

5. `compare_models()` does not do any optimization as it fits all the models with their default hyperparameters. Therefore, `sort` has been chosen as an argument name that reflects the behavior better. In contrast, `tune_model()` attempts to find hyperparameter values for a given model that maximize/minimize the chosen metric. You can set the `optimize` parameter in `ensemble_model()`, `blend_models()` and `stack_models()`, but it only works when used alongside `choose_better=True`. `choose_better=True` will return the original model/the best of the original models if the metric selected in `optimize` is better than in the ensembled/blended/stacked model - however, neither of those methods does any optimization themselves, and if `choose_better` is set to `False` `optimize` parameter will be ignored. The only method that optimizes hyperparameters is `tune_model`.

**Comment by JianWang2016** [comment]
Thanks so much for your prompt feedback.  The answers are very helpful.

To clarify the statement in 5 below
"choose_better=True. choose_better=True will return the original model/the best of the original models if the metric selected in optimize is better than in the ensembled/blended/stacked model".

e.g. boosted_dt = ensemble_model(dt, method = 'Boosting', fold=5, optimize='AUC', choose_better=True)

The comparison will be an AUC from one of the 10 estimators and the mean AUC based on the ensemble result, correct? I haven't yet to see an example where an original would be better, how can I tell from the model output which is returned, original one or ensemble result?

Does this ensemble perform the decision tree prediction (subsampling and build the tree) for each of the estimators when ensembl_model is called or it just takes the object dt and move onto boosting?

**Comment by Yard1** [comment]
`choose_better` compares the mean score for all the folds of the original model (`dt`) with the final ensembled model.

You can tell that the original model was returned by comparing it with `dt`:
```python
is_original_model = boosted_dt is dt
print(is_original_model) # will print True or False
```

Model ensembling is done through sklearn's `BaggingClassifier`/`BaggingRegressor` if method is set to `Bagging` and `AdaBoostClassifier`/`AdaBoostRegressor` if method is set to `Boosting`. Please refer to sklearn's documentation for more information.

**Comment by JianWang2016** [comment]
thanks much.
